src/test/java/com/commercetools/sync/services/impl/ProductTypeServiceImplTest.java

```diff
@@ -0,0 +1,79 @@
+package com.commercetools.sync.services.impl;
+
+import com.commercetools.sync.producttypes.ProductTypeSyncOptions;
+import com.commercetools.sync.producttypes.ProductTypeSyncOptionsBuilder;
+import com.commercetools.sync.services.ProductTypeService;
+import io.sphere.sdk.client.BadGatewayException;
+import io.sphere.sdk.client.SphereClient;
+import io.sphere.sdk.producttypes.ProductType;
+import io.sphere.sdk.producttypes.queries.ProductTypeQuery;
+import io.sphere.sdk.utils.CompletableFutureUtils;
+import org.junit.Test;
+
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class ProductTypeServiceImplTest {
+
+    @Test
+    public void fetchProductType_WithEmptyKey_ShouldNotFetchCategory() {
+        // preparation
+        final SphereClient sphereClient = mock(SphereClient.class);
+        final ProductTypeSyncOptions syncOptions = ProductTypeSyncOptionsBuilder
+            .of(sphereClient)
+            .build();
+        final ProductTypeService productTypeService = new ProductTypeServiceImpl(syncOptions);
+
+        // test
+        final CompletionStage<Optional<ProductType>> result = productTypeService.fetchProductType("");
+
+        // assertions
+        assertThat(result).isCompletedWithValue(Optional.empty());
+        verify(sphereClient, never()).execute(any());
+    }
+
+    @Test
+    public void fetchProductType_WithNullKey_ShouldNotFetchCategory() {
+        // preparation
+        final SphereClient sphereClient = mock(SphereClient.class);
+        final ProductTypeSyncOptions syncOptions = ProductTypeSyncOptionsBuilder
+            .of(sphereClient)
+            .build();
+        final ProductTypeService productTypeService = new ProductTypeServiceImpl(syncOptions);
+
+        // test
+        final CompletionStage<Optional<ProductType>> result = productTypeService.fetchProductType(null);
+
+        // assertions
+        assertThat(result).isCompletedWithValue(Optional.empty());
+        verify(sphereClient, never()).execute(any());
+    }
+
+    @Test
+    public void fetchProductType_WithBadGateWayException_ShouldCompleteExceptionally() {
```

Comment:
What about a test to check that `keyToIdCache` was filled with product type id? It is not so important?

Reply:
@lojzatran good catch, but I already know about and included it in this issue: https://github.com/commercetools/commercetools-sync-java/issues/235 (check the last point)  😉  It will be done along with the refactoring of caching later on.